lib/stores/S3Store.ts

```diff
@@ -234,40 +295,58 @@ export default class S3Store extends DataStore {
     readStream: http.IncomingMessage | fs.ReadStream,
     currentPartNumber: number,
     offset: number
-  ): Promise<void> {
+  ): Promise<number> {
     const size = metadata.file.size as number
-    const promises: Promise<void | string>[] = []
+    const promises: Promise<void>[] = []
     let pendingChunkFilepath: string | null = null
+    let bytesUploaded = 0
     const splitterStream = new StreamSplitter({
       chunkSize: this.calcOptimalPartSize(size),
       directory: os.tmpdir(),
     })
       .on('chunkStarted', (filepath) => {
         pendingChunkFilepath = filepath
       })
-      .on('chunkFinished', ({path, size: chunkSize}) => {
+      .on('chunkFinished', ({path, size: partSize}) => {
         pendingChunkFilepath = null
-        offset += chunkSize
-        const partNumber = currentPartNumber++
-        const p = Promise.resolve()
-          .then(() => {
-            // Skip chunk if it is not last and is smaller than 5MB
-            const is_last_chunk = size === offset
-            if (!is_last_chunk && chunkSize < 5 * 1024 * 1024) {
-              log(`[${metadata.file.id}] ignoring chuck smaller than 5MB`)
-              return
+        const deferred = pDefer<void>()
+
+        // Fork off in the background
+        ;(async () => {
+          try {
+            const partNumber = currentPartNumber++
+            const incompletePartId = this.partKey(metadata.file.id, true)
+            // If we received a chunk under the minimum part size in a previous iteration,
+            // we used a regular S3 upload to save it in the bucket. We try to get the incomplete part here.
+            const incompletePart = await this.getIncompletePart(incompletePartId)
+            const isFinalChunk = size === offset + partSize
```

Comment:
Maybe we need a check here if the upload length is not deferred.

Reply:
It's not possible to reach this code for deferred files, `declareUploadLength` would be called before we hit `store.write`:

https://github.com/tus/tus-node-server/blob/accb18b2f53d113ded2e176213f353ccc4de8f48/lib/handlers/PatchHandler.ts#L76-L80

Reply:
Ok, great!